internal/state/store_test.go

```diff
@@ -307,3 +306,67 @@ func TestABCIResponsesResultsHash(t *testing.T) {
 	require.NoError(t, err)
 	assert.NoError(t, proof.Verify(root, bz))
 }
+
+func TestLastABCIResponses(t *testing.T) {
```

Comment:
It looks like the pieces of this function are effectively different little tests. I would recommend using go subtests for this. Each unit of this test that represents a unique case to test can be surrounded with 
```
t.Run(<name>, func(t *testing.T) {

})
```

As an example, see https://github.com/tendermint/tendermint/blob/c287a94153572c9dc708303ff8502568450686ef/types/tx_test.go#L51

This helps keep track of all of the resources used for an individual test by creating a unique scope for each. This can be helpful in avoiding accidentally re-using a resource you didn't intend to.